### PR TITLE
Product single refactor

### PR DESCRIPTION
**Pull request by justinemorgan-11**
Refactored code for ProductSingle to get data from the store, added reviews

### CODE REVIEW

File: client/components/ProductSingle.js [hunk 1]
```diff
@@ -43,15 +43,25 @@ class ProductSingle extends React.Component {
...
-          <button type="submit">Return to Main Menu</button>
-        </Link>
-        {reviews.map(review => {
-          return <div key={review.id}>{review.comments}</div>;
-        })}
+        <div>
+          <h1>
+            <i>Reviews</i>
+          </h1>
+          {reviews.map(review => {
```
Comment (theo333) [inline client/components/ProductSingle.js:70]:
```suggestion
          {reviews.map(review => {
          const { id, rating, comment } = review;
```

File: client/components/ProductSingle.js [hunk 2]
```diff
@@ -43,15 +43,25 @@ class ProductSingle extends React.Component {
...
+            return (
+              <ul key={review.id}>
```
Comment (theo333) [inline client/components/ProductSingle.js]:
```suggestion
              <ul key={id}>
```

File: client/components/ProductSingle.js [hunk 3]
```diff
@@ -43,15 +43,25 @@ class ProductSingle extends React.Component {
...
+                <li>{review.rating}</li>
```
Comment (theo333) [inline client/components/ProductSingle.js]:
```suggestion
                <li>{rating}</li>
```

File: client/components/ProductSingle.js [hunk 4]
```diff
@@ -43,15 +43,25 @@ class ProductSingle extends React.Component {
...
+                <li>{review.comment}</li>
```
Comment (theo333) [inline client/components/ProductSingle.js]:
```suggestion
                <li>{comment}</li>
```

File: client/store.js [hunk 5]
```diff
@@ -25,15 +26,28 @@ const setCategories = categories => ({
...
+      .then(products => dispatch(setProducts(products)));
+  };
+};
+
+export const fetchProduct = id => {
+  return dispatch => {
+    return axios
+      .get(`api/products/${id}`)
+      .then(res => res.data)
+      .then(product => dispatch(setProducts([product])));
```
Comment (theo333) [inline client/store.js]: I think best practice to use an action type and action creator for only one purpose and not try to reuse a plural products for singular product.  That could get confusing.

Create an action type and action creator for each type:
```
const SET_PRODUCT = 'SET_PRODUCT';

const setProduct = product => ({ type: SET_PRODUCT, product });
```
Also add a reducer, `product`.  We will also be using this reducer later for adding and deleting a product.
 

```suggestion
      .then(product => dispatch(setProducts(product)));
```

File: client/components/ProductSingle.js [hunk 6]
```diff
@@ -1,37 +1,37 @@
...
-    };
-  }
-
-  componentDidMount() {
-    // TODO: what's a more efficient way of doing this (that linter likes)?
-    const { match } = this.props;
-    const { params } = match;
-    const { productId } = params;
+const mapStateToProps = (state, props) => {
+  const { products } = state.product;
```
Comment (theo333) [inline client/components/ProductSingle.js]: See my comments in `store` about creating a `product` reducer, along with action type and action creator.

```suggestion
  const { product } = state.product;
```

File: client/components/ProductSingle.js [hunk 7]
```diff
@@ -1,37 +1,37 @@
...
+  const { reviews } = state.review;
+  return {
+    products,
```
Comment (theo333) [inline client/components/ProductSingle.js]:
```suggestion
    product,
```

File: client/store.js [hunk 8]
```diff
@@ -69,20 +92,10 @@ export const filterProducts = categoryIds => {
...
-      .then(product => dispatch(setProducts(product)));
-  };
-};
-
 export const checkUser = enteredUser => async dispatch => {
   try {
-    const response = await axios.put('/api/auth/login', enteredUser);
-    const user = response.data;
-    return dispatch(setUser(user));
+    await axios.delete('/api/auth/logout');
```
Comment (theo333) [inline client/store.js]: Why was this changed to `delete`?  Isn't this something Erin is working on?

File: client/store.js [hunk 9]
```diff
@@ -69,20 +92,10 @@ export const filterProducts = categoryIds => {
...
+    return dispatch(setUser({}));
```
Comment (theo333) [inline client/store.js]: ? (see line above comment)

File: client/store.js [hunk 10]
```diff
@@ -134,10 +147,20 @@ const category = (state = { categories: [] }, action) => {
   }
 };
 
+const review = (state = { reviews: [] }, action) => {
```
Comment (theo333) [inline client/store.js]: The reducers were not originally set up with the correct syntax.  I have the correct one on my branch.

When we have separate reducers that we combine, the name of the reducer (e.g. `review`) becomes the property key in state.  So in this case it would be `state.review`, so no need to add a key in the initial state.

```suggestion
const reviews = (state = [], action) => {
```

File: client/store.js [hunk 11]
```diff
@@ -134,10 +147,20 @@ const category = (state = { categories: [] }, action) => {
...
+  switch (action.type) {
+    case SET_REVIEWS:
+      return { ...state, reviews: action.reviews };
```
Comment (theo333) [inline client/store.js]: Here we are setting all of the reviews from the db onto state.  (Yes, I know, looks very different than when have one reducer.  I did this for the other reducers and now state is showing properly.)
```suggestion
      return action.reviews ;
```

File: client/store.js [hunk 12]
```diff
@@ -134,10 +147,20 @@ const category = (state = { categories: [] }, action) => {
...
+    default:
+      return state;
+  }
+};
+
 const reducer = combineReducers({
   product,
   user,
   category,
+  review,
```
Comment (theo333) [inline client/store.js]:
```suggestion
  reviews,
```

File: package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "grace-shopper-funky-monkey",
-  "version": "0.3.0",
+  "version": "0.5.0",
```
Comment (theo333) [inline package.json]: Going to merge Erin's 2 PR's first, your version should change.

```suggestion
  "version": "0.7.0",
```

File: client/components/ProductSingle.js [hunk 14]
```diff
@@ -1,37 +1,37 @@
...
+    const { productId } = this.props.match.params;
+    const { fetchProduct, fetchProductReviews } = this.props;
+    fetchProductReviews(productId);
+    fetchProduct(productId);
+  };
 
   render() {
-    const { product, reviews } = this.state;
-
+    const { products, reviews } = this.props;
```
Comment (theo333) [inline client/components/ProductSingle.js]: See my comments in store about creating a product reducer, along with action type and action creator.
```suggestion
    const { product, reviews } = this.props;
```

File: client/components/ProductSingle.js [hunk 15]
```diff
@@ -1,37 +1,37 @@
...
+    if (!products.length) return null;
+    const product = products[0];
```
Comment (theo333) [inline client/components/ProductSingle.js]: See my comments in store about creating a product reducer, along with action type and action creator.
```suggestion

```

File: client/components/ProductSingle.js [hunk 16]
```diff
@@ -1,37 +1,37 @@
...
+    const { fetchProduct, fetchProductReviews } = this.props;
+    fetchProductReviews(productId);
+    fetchProduct(productId);
+  };
 
   render() {
-    const { product, reviews } = this.state;
-
+    const { products, reviews } = this.props;
+    if (!products.length) return null;
```
Comment (theo333) [inline client/components/ProductSingle.js]: See my comments in store about creating a product reducer, along with action type and action creator.

Why would this ever be the case that it is empty?  If it is empty, then should never get to this component since are coming to this route via `<Route exact path="/menu/:productId" component={ProductSingle} />`.  Correct?

So probably don't need this line at all.
```suggestion
    if (!product.length) return null;
```

**Review by theo333 — CHANGES_REQUESTED** [review]
Overall good.  There are some things in here that are confusing, like re-using plural `products` to show a single `product`.  I think not a good practice.  

We are going to merge your PR after Erin's since she is busy at her conference and will not have the time to rebase her code.  Hopefully you will have the time.  So this also means changing the version for this PR in package.json.  See my comment.

I hope my comments make sense.  If it is easy to video chat (and you are able to from work), then I am happy to do that and work together to make the changes.

Reviews - It would be good to have a first name displayed with a review.  I just added an Asana task for this for Stage 2.

File: client/components/ProductSingle.js (hunk 16 above)
Comment (justinemorgan-11) [reply client/components/ProductSingle.js]: When I go directly to the page, it tries to render once with no data then renders again once it's got the data. I was getting "undefined" errors before adding this if statement.

File: client/store.js (hunk 8 above)
Comment (justinemorgan-11) [reply client/store.js]: Shouldn't be, manually changing it back to whatever Erin currently has in dev.

File: client/store.js (hunk 9 above)
Comment (justinemorgan-11) [reply client/store.js]: Same as above.

**Review by theo333 — APPROVED** [review]
